File: backend/src/routes/trips.ts
import express from 'express';
import { Trip } from '../models/Trip.js';
import { authenticateToken } from '../middleware/auth.js';

const router = express.Router();

router.use(authenticateToken);

router.post('/', async (req, res) => {
  try {
    const {
      title, description, startDate, startLocation, cities,
      totalDays, people, travelType, budget, budgetMode,
      generatedItinerary, travelMeans, isPublic = false, tags = [],
    } = req.body;

    if (!title || !startDate || !cities || !people || !travelType || !budget || !generatedItinerary) {
      return res.status(400).json({
        error: 'Missing required fields: title, startDate, cities, people, travelType, budget, generatedItinerary',
      });
    }

    const trip = new Trip({
      user: req.userId,
      title, description,
      startDate: new Date(startDate),
      startLocation, cities, totalDays, people, travelType,
      budget, budgetMode, generatedItinerary, travelMeans,
      isPublic, tags,
    });

    await trip.save();
    return res.status(201).json({ message: 'Trip saved successfully', savedTrip: trip });
  } catch (error: any) {
    console.error('Error saving trip:', error);
    return res.status(500).json({ error: 'Failed to save trip', details: error.message });
  }
});

router.get('/', async (req, res) => {
  try {
    const { page = 1, limit = 10, search } = req.query;
    const skip = (Number(page) - 1) * Number(limit);

    const query: any = {
      user: req.userId,
      isUpcoming: { $ne: true },
      isCompleted: { $ne: true },
    };

    if (search) {
      query.$and = [{
        $or: [
          { title: { $regex: search, $options: 'i' } },
          { description: { $regex: search, $options: 'i' } },
          { tags: { $in: [new RegExp(search as string, 'i')] } },
        ],
      }];
    }

    const savedTrips = await Trip.find(query)
      .sort({ createdAt: -1 })
      .skip(skip)
      .limit(Number(limit));

    const total = await Trip.countDocuments(query);

    return res.json({
      savedTrips,
      pagination: {
        current: Number(page),
        pages: Math.ceil(total / Number(limit)),
        total,
      },
    });
  } catch (error: any) {
    console.error('Error fetching saved trips:', error);
    return res.status(500).json({ error: 'Failed to fetch saved trips', details: error.message });
  }
});

router.get('/upcoming', async (req, res) => {
  try {
    const { page = 1, limit = 10 } = req.query;
    const skip = (Number(page) - 1) * Number(limit);

    const upcomingTrips = await Trip.find({
      user: req.userId,
      isUpcoming: true,
      isCompleted: { $ne: true },
    })
      .sort({ tripStartDate: 1 })
      .skip(skip)
      .limit(Number(limit));

    const total = await Trip.countDocuments({
      user: req.userId,
      isUpcoming: true,
      isCompleted: { $ne: true },
    });

    return res.json({
      upcomingTrips,
      pagination: { current: Number(page), pages: Math.ceil(total / Number(limit)), total },
    });
  } catch (error: any) {
    return res.status(500).json({ error: 'Failed to fetch upcoming trips', details: error.message });
  }
});

router.get('/completed', async (req, res) => {
  try {
    const { page = 1, limit = 10 } = req.query;
    const skip = (Number(page) - 1) * Number(limit);

    const completedTrips = await Trip.find({
      user: req.userId,
      isCompleted: true,
    })
      .sort({ tripEndDate: -1 })
      .skip(skip)
      .limit(Number(limit));

    const total = await Trip.countDocuments({
      user: req.userId,
      isCompleted: true,
    });

    return res.json({
      completedTrips,
      pagination: { current: Number(page), pages: Math.ceil(total / Number(limit)), total },
    });
  } catch (error: any) {
    return res.status(500).json({ error: 'Failed to fetch completed trips', details: error.message });
  }
});

router.get('/statistics', async (req, res) => {
  try {
    const allTrips = await Trip.find({ user: req.userId });

    const totalTrips = allTrips.length;
    const savedTrips = allTrips.filter((t) => t.isUpcoming !== true && t.isCompleted !== true).length;
    const upcomingTripsCount = allTrips.filter((t) => t.isUpcoming === true && t.isCompleted !== true).length;
    const completedTripsCount = allTrips.filter((t) => t.isCompleted === true).length;

    const completedTripsList = allTrips.filter((t) => t.isCompleted === true);
    const totalDaysTraveled = completedTripsList.reduce((sum, t) => sum + (t.totalDays || 0), 0);

    const citiesVisited = new Set<string>();
    completedTripsList.forEach((t) => t.cities?.forEach((c) => citiesVisited.add(c.name)));

    const countriesVisited = new Set<string>();
    completedTripsList.forEach((t) => t.cities?.forEach((c) => {
      const parts = c.name.split(',');
      if (parts.length > 1) countriesVisited.add(parts[parts.length - 1].trim());
    }));

    const totalActivities = allTrips.reduce((sum, t) => {
      if (!t.generatedItinerary?.days) return sum;
      return sum + t.generatedItinerary.days.reduce(
        (daySum, day) => daySum + (day.timeSlots || []).reduce(
          (slotSum, slot) => slotSum + (slot.activities || []).length, 0
        ), 0
      );
    }, 0);

    const recentTrip = completedTripsList
      .sort((a, b) => new Date(b.tripEndDate || 0).getTime() - new Date(a.tripEndDate || 0).getTime())[0];

    const nextTrip = allTrips
      .filter((t) => t.isUpcoming === true && t.isCompleted !== true)
      .sort((a, b) => new Date(a.tripStartDate || 0).getTime() - new Date(b.tripStartDate || 0).getTime())[0];

    return res.json({
      statistics: {
        totalTrips, savedTrips,
        upcomingTrips: upcomingTripsCount,
        completedTrips: completedTripsCount,
        totalDaysTraveled,
        citiesVisited: citiesVisited.size,
        countriesVisited: countriesVisited.size,
        totalActivities,
        recentTrip: recentTrip ? {
          id: recentTrip._id, title: recentTrip.title,
          cities: recentTrip.cities, endDate: recentTrip.tripEndDate,
        } : null,
        nextTrip: nextTrip ? {
          id: nextTrip._id, title: nextTrip.title,
          cities: nextTrip.cities, startDate: nextTrip.tripStartDate,
        } : null,
      },
    });
  } catch (error: any) {
    return res.status(500).json({ error: 'Failed to fetch trip statistics', details: error.message });
  }
});

router.get('/check', async (req, res) => {
  try {
    const { startDate, cities, people, travelType } = req.query;

    if (!startDate || !cities || !people || !travelType) {
      return res.status(400).json({ error: 'Missing required query parameters' });
    }

    const existingTrip = await Trip.findOne({
      user: req.userId,
      startDate: new Date(startDate as string),
      people: parseInt(people as string),
      travelType: travelType as string,
      'cities.name': { $in: JSON.parse(cities as string).map((c: any) => c.name) },
    });

    return res.json({ isSaved: !!existingTrip, savedTrip: existingTrip });
  } catch (error: any) {
    return res.status(500).json({ error: 'Failed to check trip status', details: error.message });
  }
});

router.get('/:id', async (req, res) => {
  try {
    const trip = await Trip.findOne({ _id: req.params.id, user: req.userId });
    if (!trip) return res.status(404).json({ error: 'Saved trip not found' });
    return res.json(trip);
  } catch (error: any) {
    return res.status(500).json({ error: 'Failed to fetch saved trip', details: error.message });
  }
});

router.put('/:id', async (req, res) => {
  try {
    const updates = { ...req.body };
    delete updates.user;
    delete updates._id;
    delete updates.createdAt;

    const trip = await Trip.findOneAndUpdate(
      { _id: req.params.id, user: req.userId },
      updates,
      { new: true, runValidators: true }
    );

    if (!trip) return res.status(404).json({ error: 'Saved trip not found' });
    return res.json({ message: 'Trip updated successfully', savedTrip: trip });
  } catch (error: any) {
    return res.status(500).json({ error: 'Failed to update saved trip', details: error.message });
  }
});

router.put('/:id/upcoming', async (req, res) => {
  try {
    const { tripStartDate } = req.body;
    if (!tripStartDate) return res.status(400).json({ error: 'Trip start date is required' });

    const trip = await Trip.findOne({ _id: req.params.id, user: req.userId });
    if (!trip) return res.status(404).json({ error: 'Saved trip not found' });

    const startDate = new Date(tripStartDate);
    const endDate = new Date(startDate);
    endDate.setDate(endDate.getDate() + trip.totalDays);

    const updated = await Trip.findOneAndUpdate(
      { _id: req.params.id, user: req.userId },
      { isUpcoming: true, isCompleted: false, tripStartDate: startDate, tripEndDate: endDate },
      { new: true, runValidators: true }
    );

    return res.json({ message: 'Trip marked as upcoming successfully', savedTrip: updated });
  } catch (error: any) {
    return res.status(500).json({ error: 'Failed to mark trip as upcoming', details: error.message });
  }
});

router.put('/:id/completed', async (req, res) => {
  try {
    const trip = await Trip.findOneAndUpdate(
      { _id: req.params.id, user: req.userId },
      { isCompleted: true, tripEndDate: new Date() },
      { new: true, runValidators: true }
    );

    if (!trip) return res.status(404).json({ error: 'Saved trip not found' });
    return res.json({ message: 'Trip marked as completed successfully', savedTrip: trip });
  } catch (error: any) {
    return res.status(500).json({ error: 'Failed to mark trip as completed', details: error.message });
  }
});

router.delete('/:id/upcoming', async (req, res) => {
  try {
    const trip = await Trip.findOneAndUpdate(
      { _id: req.params.id, user: req.userId },
      { isUpcoming: false, isCompleted: false, tripStartDate: null, tripEndDate: null },
      { new: true, runValidators: true }
    );

    if (!trip) return res.status(404).json({ error: 'Saved trip not found' });
    return res.json({ message: 'Trip moved back to saved successfully', savedTrip: trip });
  } catch (error: any) {
    return res.status(500).json({ error: 'Failed to remove trip from upcoming', details: error.message });
  }
});

router.delete('/:id', async (req, res) => {
  try {
    const trip = await Trip.findOneAndDelete({ _id: req.params.id, user: req.userId });
    if (!trip) return res.status(404).json({ error: 'Saved trip not found' });
    return res.json({ message: 'Trip deleted successfully' });
  } catch (error: any) {
    return res.status(500).json({ error: 'Failed to delete saved trip', details: error.message });
  }
});

export default router;

File: backend-python/app/routes/trips.py
"""Trips routes — full CRUD + statistics + upcoming/completed."""

from datetime import datetime, timedelta
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy import select, func, or_
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.deps import get_current_user
from app.models import Trip, User
from app.schemas.trip import CreateTripRequest, UpdateTripRequest, MarkUpcomingRequest

router = APIRouter()


def _trip_to_dict(trip: Trip) -> dict:
    return {
        "id": trip.id,
        "title": trip.title,
        "description": trip.description,
        "startDate": trip.start_date.isoformat() if trip.start_date else None,
        "startLocation": trip.start_location,
        "cities": trip.cities or [],
        "totalDays": trip.total_days,
        "people": trip.people,
        "travelType": trip.travel_type,
        "budget": trip.budget,
        "budgetMode": trip.budget_mode,
        "generatedItinerary": trip.generated_itinerary,
        "isPublic": trip.is_public,
        "tags": trip.tags or [],
        "isUpcoming": trip.is_upcoming,
        "isCompleted": trip.is_completed,
        "tripStartDate": trip.trip_start_date.isoformat() if trip.trip_start_date else None,
        "tripEndDate": trip.trip_end_date.isoformat() if trip.trip_end_date else None,
        "createdAt": trip.created_at.isoformat() if trip.created_at else None,
        "updatedAt": trip.updated_at.isoformat() if trip.updated_at else None,
    }


@router.post("")
@router.post("/")
async def create_trip(
    req: CreateTripRequest,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    trip = Trip(
        user_id=user.id,
        title=req.title,
        description=req.description,
        start_date=datetime.fromisoformat(req.startDate) if req.startDate else None,
        start_location=req.startLocation,
        cities=[c.model_dump() for c in req.cities],
        total_days=req.totalDays,
        people=req.people,
        travel_type=req.travelType,
        budget=req.budget.model_dump() if req.budget else None,
        budget_mode=req.budgetMode,
        generated_itinerary=req.generatedItinerary,
        is_public=req.isPublic,
        tags=req.tags,
    )
    db.add(trip)
    await db.commit()
    await db.refresh(trip)
    return {"message": "Trip saved successfully", "savedTrip": _trip_to_dict(trip)}


@router.get("")
@router.get("/")
async def list_trips(
    page: int = 1,
    limit: int = 10,
    search: str | None = None,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    query = select(Trip).where(
        Trip.user_id == user.id,
        Trip.is_upcoming != True,
        Trip.is_completed != True,
    )
    if search:
        query = query.where(
            or_(
                Trip.title.ilike(f"%{search}%"),
                Trip.description.ilike(f"%{search}%"),
            )
        )
    query = query.order_by(Trip.created_at.desc()).offset((page - 1) * limit).limit(limit)
    result = await db.execute(query)
    trips = result.scalars().all()

    count_q = select(func.count(Trip.id)).where(
        Trip.user_id == user.id,
        Trip.is_upcoming != True,
        Trip.is_completed != True,
    )
    total = (await db.execute(count_q)).scalar() or 0

    return {
        "savedTrips": [_trip_to_dict(t) for t in trips],
        "pagination": {
            "current": page,
            "pages": (total + limit - 1) // limit,
            "total": total,
        },
    }


@router.get("/upcoming")
async def list_upcoming(
    page: int = 1,
    limit: int = 10,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    query = select(Trip).where(
        Trip.user_id == user.id,
        Trip.is_upcoming == True,
        Trip.is_completed != True,
    ).order_by(Trip.trip_start_date.asc()).offset((page - 1) * limit).limit(limit)
    result = await db.execute(query)
    trips = result.scalars().all()

    count_q = select(func.count(Trip.id)).where(
        Trip.user_id == user.id,
        Trip.is_upcoming == True,
        Trip.is_completed != True,
    )
    total = (await db.execute(count_q)).scalar() or 0

    return {
        "upcomingTrips": [_trip_to_dict(t) for t in trips],
        "pagination": {"current": page, "pages": (total + limit - 1) // limit, "total": total},
    }


@router.get("/completed")
async def list_completed(
    page: int = 1,
    limit: int = 10,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    query = select(Trip).where(
        Trip.user_id == user.id,
        Trip.is_completed == True,
    ).order_by(Trip.trip_end_date.desc()).offset((page - 1) * limit).limit(limit)
    result = await db.execute(query)
    trips = result.scalars().all()

    count_q = select(func.count(Trip.id)).where(
        Trip.user_id == user.id,
        Trip.is_completed == True,
    )
    total = (await db.execute(count_q)).scalar() or 0

    return {
        "completedTrips": [_trip_to_dict(t) for t in trips],
        "pagination": {"current": page, "pages": (total + limit - 1) // limit, "total": total},
    }


@router.get("/statistics")
async def get_statistics(
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    result = await db.execute(select(Trip).where(Trip.user_id == user.id))
    all_trips = result.scalars().all()

    saved = [t for t in all_trips if not t.is_upcoming and not t.is_completed]
    upcoming = [t for t in all_trips if t.is_upcoming and not t.is_completed]
    completed = [t for t in all_trips if t.is_completed]

    cities_visited = set()
    countries_visited = set()
    for t in completed:
        for c in (t.cities or []):
            if isinstance(c, dict):
                cities_visited.add(c.get("name", ""))
                parts = c.get("name", "").split(",")
                if len(parts) > 1:
                    countries_visited.add(parts[-1].strip())

    total_days = sum(t.total_days or 0 for t in completed)
    total_activities = 0
    for t in all_trips:
        days = (t.generated_itinerary or {}).get("days", [])
        for day in days:
            for slot in day.get("timeSlots", []):
                total_activities += len(slot.get("activities", []))

    recent = sorted(completed, key=lambda t: t.trip_end_date or datetime.min, reverse=True)
    next_trip = sorted(upcoming, key=lambda t: t.trip_start_date or datetime.max)

    def _brief(t):
        return {"id": t.id, "title": t.title, "cities": t.cities or []}

    return {
        "statistics": {
            "totalTrips": len(all_trips),
            "savedTrips": len(saved),
            "upcomingTrips": len(upcoming),
            "completedTrips": len(completed),
            "totalDaysTraveled": total_days,
            "citiesVisited": len(cities_visited),
            "countriesVisited": len(countries_visited),
            "totalActivities": total_activities,
            "recentTrip": {**_brief(recent[0]), "endDate": recent[0].trip_end_date.isoformat() if recent[0].trip_end_date else None} if recent else None,
            "nextTrip": {**_brief(next_trip[0]), "startDate": next_trip[0].trip_start_date.isoformat() if next_trip[0].trip_start_date else None} if next_trip else None,
        }
    }


@router.get("/check")
async def check_trip(
    startDate: str = Query(...),
    cities: str = Query(...),
    people: int = Query(...),
    travelType: str = Query(...),
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    import json
    city_names = [c["name"] for c in json.loads(cities)]
    result = await db.execute(
        select(Trip).where(
            Trip.user_id == user.id,
            Trip.start_date == datetime.fromisoformat(startDate),
            Trip.people == people,
            Trip.travel_type == travelType,
        )
    )
    trips = result.scalars().all()
    for trip in trips:
        trip_city_names = [c.get("name") for c in (trip.cities or [])]
        if any(cn in trip_city_names for cn in city_names):
            return {"isSaved": True, "savedTrip": _trip_to_dict(trip)}
    return {"isSaved": False, "savedTrip": None}


@router.get("/{trip_id}")
async def get_trip(
    trip_id: int,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    result = await db.execute(select(Trip).where(Trip.id == trip_id, Trip.user_id == user.id))
    trip = result.scalar_one_or_none()
    if not trip:
        raise HTTPException(status_code=404, detail="Saved trip not found")
    return _trip_to_dict(trip)


@router.put("/{trip_id}")
async def update_trip(
    trip_id: int,
    req: UpdateTripRequest,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    result = await db.execute(select(Trip).where(Trip.id == trip_id, Trip.user_id == user.id))
    trip = result.scalar_one_or_none()
    if not trip:
        raise HTTPException(status_code=404, detail="Saved trip not found")

    data = req.model_dump(exclude_unset=True)
    for key, value in data.items():
        if key == "startDate" and value:
            trip.start_date = datetime.fromisoformat(value)
        elif key == "cities" and value:
            trip.cities = value
        elif key == "budget" and value:
            trip.budget = value
        elif hasattr(trip, key):
            setattr(trip, key, value)

    await db.commit()
    await db.refresh(trip)
    return {"message": "Trip updated successfully", "savedTrip": _trip_to_dict(trip)}


@router.put("/{trip_id}/upcoming")
async def mark_upcoming(
    trip_id: int,
    req: MarkUpcomingRequest,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    result = await db.execute(select(Trip).where(Trip.id == trip_id, Trip.user_id == user.id))
    trip = result.scalar_one_or_none()
    if not trip:
        raise HTTPException(status_code=404, detail="Saved trip not found")

    start = datetime.fromisoformat(req.tripStartDate)
    end = start + timedelta(days=trip.total_days or 1)
    trip.is_upcoming = True
    trip.is_completed = False
    trip.trip_start_date = start.date()
    trip.trip_end_date = end.date()
    await db.commit()
    await db.refresh(trip)
    return {"message": "Trip marked as upcoming successfully", "savedTrip": _trip_to_dict(trip)}


@router.put("/{trip_id}/completed")
async def mark_completed(
    trip_id: int,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    result = await db.execute(select(Trip).where(Trip.id == trip_id, Trip.user_id == user.id))
    trip = result.scalar_one_or_none()
    if not trip:
        raise HTTPException(status_code=404, detail="Saved trip not found")
    trip.is_completed = True
    trip.trip_end_date = datetime.utcnow().date()
    await db.commit()
    await db.refresh(trip)
    return {"message": "Trip marked as completed successfully", "savedTrip": _trip_to_dict(trip)}


@router.delete("/{trip_id}/upcoming")
async def remove_upcoming(
    trip_id: int,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    result = await db.execute(select(Trip).where(Trip.id == trip_id, Trip.user_id == user.id))
    trip = result.scalar_one_or_none()
    if not trip:
        raise HTTPException(status_code=404, detail="Saved trip not found")
    trip.is_upcoming = False
    trip.is_completed = False
    trip.trip_start_date = None
    trip.trip_end_date = None
    await db.commit()
    await db.refresh(trip)
    return {"message": "Trip moved back to saved successfully", "savedTrip": _trip_to_dict(trip)}


@router.delete("/{trip_id}")
async def delete_trip(
    trip_id: int,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    result = await db.execute(select(Trip).where(Trip.id == trip_id, Trip.user_id == user.id))
    trip = result.scalar_one_or_none()
    if not trip:
        raise HTTPException(status_code=404, detail="Saved trip not found")
    await db.delete(trip)
    await db.commit()
    return {"message": "Trip deleted successfully"}
